### DISCUSSION TITLE
Async Arduino Nano only responding to first request.

### DISCUSSION THREAD

**Discussion by Tavaru5** [miguelgrinberg/microdot#276] [Q&A]
I have a similar issue to https://github.com/miguelgrinberg/microdot/discussions/166 / https://github.com/miguelgrinberg/microdot/issues/89, where my server seems to be hanging after the first request/response. My issue is that I'm already using async (or at least I believe I am), so I'm not sure why I would still be having this issue. I'm using micropython, then testing it with postman. It's loosely based off of the [gpio example](https://github.com/miguelgrinberg/microdot/tree/main/examples/gpio), but slimmed down a lot.

My code looks like this:
```
import network
import time
import asyncio

from machine import Pin
from microdot import Microdot

# { network initialization code goes here }

app = Microdot()

@app.route('/pin', methods=['GET', 'POST'])
async def handlePin(request):
  if ('pin' in request.json):
    pinId = request.json['pin']
    if (request.method == 'POST'):
      if ('value' in request.json):
        value = request.json['value']
        pin = Pin(pinId, Pin.OUT)
        if (value == 'ON'):
          pin.on()
        elif (value == 'OFF'):
          pin.off()
    response = dict({ 'success': True})
    pin = Pin(pinId, Pin.IN)
    response['pin'] = pinId
    response['value'] = pin.value()
    return (response)
  else:
    return { 'success': False}

async def main():
  server = asyncio.create_task(app.start_server(debug=True))
  await server

asyncio.run(main())
```

Any insight on why this is happening would be appreciated, I'm not sure if there's something obvious that I'm missing since I'm not super familiar with python.

**Comment by miguelgrinberg** [comment]
You are using async. The issues you linked are from a time when Microdot supported sync natively and async through an extension. In the current version the internals were redesigned so that the core is async, and sync options can only be used in platforms that have threads such as CPython (none of which applies to your board).
 
Questions:
- Does the client get the response for that first/only request that goes through?
- Since you are running with `debug=True`, you should see the request logged in the console (if you are connected to it via serial port). Can you check if the request appears?
- What client are you using? It may be useful to try this with a simple client such as `curl`, which does not have any of the browser behaviors I describe in the linked issues.

**Reply by Tavaru5** [reply]
Yes the client gets the response & the request is logged for the first request. I am using Postman for my requests, but when I attempt it with `curl` I get the same issue; the first request works, but after that it just hangs.

**Reply by miguelgrinberg** [reply]
Please try any of the examples in this repository, without changing anything besides adding the network connection. Do my examples also hang after one request is handled? I do not have an Arduino board unfortunately, so I cannot test this right now.

**Reply by Tavaru5** [reply]
They also hang, which I didn't expect since I thought I had tested them before. Good call to check that. 

I assume it's something with my board then (or I guess possibly something with my network initialization, although I'm not sure exactly how that would cause it).

**Reply by miguelgrinberg** [reply]
Maybe try a full reset of the board and a reflash of the MicroPython firmware. If you remember testing the examples earlier, then maybe your board got into a bad state for some reason.

Also, this is the Nano ESP32 board, correct? I may need to get one of those to add to my testing collection.

**Reply by Tavaru5** [reply]
Yeah it is the Nano ESP32. Reflashing/resetting is a good idea; I'll try that out and then see if that fixes it.

**Comment by bulletmark** [comment]
@Tavaru5 I ran your code on an ESP32 and a Pico W and it runs fine. Keeps responding to all the curl requests I send.

**Reply by Tavaru5** [reply]
Thanks for checking. It seems like at this point it's either something with my board or my network initialization.
